CMakeLists.txt

```diff
@@ -381,18 +381,33 @@ if (MSVC)
   else()
     set(EXTRA_FLAGS "")
   endif()
-  if (BUILD_SHARED_LIBS OR ONNX_BUILD_MAIN_LIB)
-    set(ONNX_API_DEFINE "-DONNX_API=__declspec(dllexport)")
+endif()
+
+# onnx_proto symbols visibility
+if(MSVC)
+  if(ONNX_BUILD_MAIN_LIB)
+  # ONNX_BUILD_MAIN_LIB can be be set if onnx_proto is static,
+  # but it is linked into a shared library which wants
+  # to export the ONNX APIs and classes.
+  target_compile_definitions(onnx_proto PUBLIC "ONNX_API=__declspec(dllexport)")
+
+  elseif(BUILD_SHARED_LIBS)
+    # users of onnx_proto import symbols
+    target_compile_definitions(onnx_proto INTERFACE "ONNX_API=__declspec(dllimport)")
+    # onnx_proto exports symbols
+    target_compile_definitions(onnx_proto PRIVATE "ONNX_API=__declspec(dllexport)")
+
   else()
-    set(ONNX_API_DEFINE "-DONNX_API=")
+    # Windows static libs don't need symbols visibility decorators
+    target_compile_definitions(onnx_proto PUBLIC "ONNX_API=")
   endif()
 else()
   # On non-Windows, hide all symbols we don't need
-  set(ONNX_API_DEFINE "-DONNX_API=__attribute__\(\(__visibility__\(\"default\"\)\)\)")
-  set_target_properties(onnx_proto PROPERTIES CXX_VISIBILITY_PRESET hidden)
-  set_target_properties(onnx_proto PROPERTIES VISIBILITY_INLINES_HIDDEN 1)
+    target_compile_definitions(onnx_proto PUBLIC "ONNX_API=__attribute__\(\(__visibility__\(\"default\"\)\)\)")
```

Comment:
what is the difference between original set vs the PR version target_compile_definitions(onnx_proto ?

is set global ?

Reply:
Previous version set value on directory level - it will be visible in current directory scope.
In the current version, we match it only with the target which uses it (it will be visible for `onnx_proto` code and for targets which use `onnx_proto`).
So from a functional point of view, it's rather not a big difference.